File: app/src/main/java/org/worshipsongs/service/PopupMenuService.java
package org.worshipsongs.service;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Paint;
import android.graphics.pdf.PdfDocument;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.print.PrintAttributes;
import android.print.pdf.PrintedPdfDocument;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.ContextThemeWrapper;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.PopupMenu;

import org.worshipsongs.CommonConstants;
import org.worshipsongs.R;
import org.worshipsongs.WorshipSongApplication;
import org.worshipsongs.activity.CustomYoutubeBoxActivity;
import org.worshipsongs.activity.PresentSongActivity;
import org.worshipsongs.dialog.FavouritesDialogFragment;
import org.worshipsongs.domain.Song;
import org.worshipsongs.utils.PermissionUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.util.Arrays;
import java.util.List;

import static android.content.Context.PRINT_SERVICE;
import static org.worshipsongs.WorshipSongApplication.getContext;

/**
 * author: Madasamy,Seenivasan, Vignesh Palanisamy
 * version: 1.0.0
 */

public class PopupMenuService
{

    private CustomTagColorService customTagColorService = new CustomTagColorService();
    private UserPreferenceSettingService preferenceSettingService = new UserPreferenceSettingService();
    private FavouriteService favouriteService = new FavouriteService();
    private SongService songService = new SongService(getContext());

    public void showPopupmenu(final AppCompatActivity activity, final View view, final String songName, boolean hidePlay)
    {
        Context wrapper = new ContextThemeWrapper(getContext(), R.style.PopupMenu_Theme);
        final PopupMenu popupMenu;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
            popupMenu = new PopupMenu(wrapper, view, Gravity.RIGHT);
        } else {
            popupMenu = new PopupMenu(wrapper, view);
        }
        popupMenu.getMenuInflater().inflate(R.menu.favourite_share_option_menu, popupMenu.getMenu());
        final Song song = songService.findContentsByTitle(songName);
        final String urlKey = song.getUrlKey();
        MenuItem menuItem = popupMenu.getMenu().findItem(R.id.play_song);
        menuItem.setVisible(urlKey != null && urlKey.length() > 0 && preferenceSettingService.isPlayVideo() && hidePlay);
        MenuItem exportMenuItem = popupMenu.getMenu().findItem(R.id.export_pdf);
        exportMenuItem.setVisible(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT);
        MenuItem presentSongMenuItem = popupMenu.getMenu().findItem(R.id.present_song);
        presentSongMenuItem.setVisible(false);
        popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener()
        {
            public boolean onMenuItemClick(final MenuItem item)
            {
                switch (item.getItemId()) {
                    case R.id.addToList:
                        Bundle bundle = new Bundle();
                        bundle.putString(CommonConstants.TITLE_KEY, songName);
                        bundle.putString(CommonConstants.LOCALISED_TITLE_KEY, song.getTamilTitle());
                        bundle.putInt(CommonConstants.ID, song.getId());
                        FavouritesDialogFragment favouritesDialogFragment = FavouritesDialogFragment.newInstance(bundle);
                        favouritesDialogFragment.show(activity.getSupportFragmentManager(), "FavouritesDialogFragment");
                        return true;
                    case R.id.share_whatsapp:
                        shareSongInSocialMedia(songName, song);
                        return true;
                    case R.id.play_song:
                        showYouTube(urlKey, songName);
                        return true;
                    case R.id.present_song:
                        startPresentActivity(songName);
                        return true;
                    case R.id.export_pdf:
                        if (PermissionUtils.isStoragePermissionGranted(activity)) {
                            exportSongToPDF(songName, Arrays.asList(song));
                        }
                        return true;
                    default:
                        return false;
                }
            }
        });
        popupMenu.show();
    }

    private void shareSongInSocialMedia(String songName, Song song)
    {
        StringBuilder builder = new StringBuilder();
        builder.append(songName).append("\n").append("\n");
        for (String content : song.getContents()) {
            builder.append(customTagColorService.getFormattedLines(content));
            builder.append("\n");
        }
        builder.append(getContext().getString(R.string.share_info));
        Log.i(PopupMenuService.this.getClass().getSimpleName(), builder.toString());
        Intent textShareIntent = new Intent(Intent.ACTION_SEND);
        textShareIntent.putExtra(Intent.EXTRA_TEXT, builder.toString());
        textShareIntent.setType("text/plain");
        Intent intent = Intent.createChooser(textShareIntent, "Share " + songName + " with...");
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    private void exportSongToPDF(String songName, List<Song> songs)
    {
        File file = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS), songName + ".pdf");

        PrintAttributes printAttrs = new PrintAttributes.Builder().
                setColorMode(PrintAttributes.COLOR_MODE_COLOR).
                setMediaSize(PrintAttributes.MediaSize.ISO_A4).
                setResolution(new PrintAttributes.Resolution("zooey", PRINT_SERVICE, 450, 700)).
                setMinMargins(PrintAttributes.Margins.NO_MARGINS).
                build();
        PdfDocument document = new PrintedPdfDocument(WorshipSongApplication.getContext(), printAttrs);
        for (int i = 0; i < songs.size(); i++) {
            Song song = songs.get(i);
            PdfDocument.PageInfo pageInfo = new PdfDocument.PageInfo.Builder(450, 700, i).create();
            PdfDocument.Page page = document.startPage(pageInfo);
            if (page != null) {
                Paint titleDesign = new Paint();
                titleDesign.setTextAlign(Paint.Align.LEFT);
                titleDesign.setTextSize(18);
                String title = song.getTamilTitle() + "/" + song.getTitle();
                float titleLength = titleDesign.measureText(title);
                float yPos = 50;
                if (page.getCanvas().getWidth() > titleLength) {
                    int xPos = (page.getCanvas().getWidth() / 2) - (int) titleLength / 2;
                    page.getCanvas().drawText(title, xPos, 20, titleDesign);
                } else {
                    int xPos = (page.getCanvas().getWidth() / 2) - (int) titleDesign.measureText(song.getTamilTitle()) / 2;
                    page.getCanvas().drawText(song.getTamilTitle() + "/", xPos, 20, titleDesign);
                    xPos = (page.getCanvas().getWidth() / 2) - (int) titleDesign.measureText(song.getTitle()) / 2;
                    page.getCanvas().drawText(song.getTitle(), xPos, 45, titleDesign);
                    yPos = 75;
                }
                for (String content : song.getContents()) {
                    if (yPos > 620) {
                        document.finishPage(page);
                        page = document.startPage(pageInfo);
                        yPos = 40;
                    }
                    yPos = customTagColorService.getFormattedPage(content, page, 10, yPos);
                    yPos = yPos + 20;
                }
            }
            document.finishPage(page);
        }
        try {
            OutputStream os = new FileOutputStream(file);
            document.writeTo(os);
            document.close();
            os.close();
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("application/pdf");
            shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.parse("file://" + file.getAbsolutePath()));

            Intent intent = Intent.createChooser(shareIntent, "Share " + songName + " with...");
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            getContext().startActivity(intent);
            Log.i("done", file.getAbsolutePath().toString());

        } catch (Exception ex) {
            Log.e(PopupMenuService.class.getSimpleName(), "Error occurred while exporting to PDF", ex);
        }
    }

    private void showYouTube(String urlKey, String songName)
    {
        Log.i(this.getClass().getSimpleName(), "Url key: " + urlKey);
        Intent youTubeIntent = new Intent(getContext(), CustomYoutubeBoxActivity.class);
        youTubeIntent.putExtra(CustomYoutubeBoxActivity.KEY_VIDEO_ID, urlKey);
        youTubeIntent.putExtra(CommonConstants.TITLE_KEY, songName);
        youTubeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(youTubeIntent);
    }

    private void startPresentActivity(String title)
    {
        Intent intent = new Intent(getContext(), PresentSongActivity.class);
        intent.putExtra(CommonConstants.TITLE_KEY, title);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
    }

    public void shareFavouritesInSocialMedia(Activity activity, View view, final String favouriteName)
    {
        Context wrapper = new ContextThemeWrapper(getContext(), R.style.PopupMenu_Theme);
        final PopupMenu popupMenu;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
            popupMenu = new PopupMenu(wrapper, view, Gravity.RIGHT);
        } else {
            popupMenu = new PopupMenu(wrapper, view);
        }
        popupMenu.getMenuInflater().inflate(R.menu.favourite_share_option_menu, popupMenu.getMenu());
        popupMenu.getMenu().findItem(R.id.play_song).setVisible(false);
        popupMenu.getMenu().findItem(R.id.present_song).setVisible(false);
        popupMenu.getMenu().findItem(R.id.addToList).setVisible(false);
        popupMenu.setOnMenuItemClickListener(getPopupMenuItemListener(activity,favouriteName));
        popupMenu.show();
    }

    @NonNull
    private PopupMenu.OnMenuItemClickListener getPopupMenuItemListener(final Activity activity, final String text)
    {
        return new PopupMenu.OnMenuItemClickListener()
        {
            public boolean onMenuItemClick(final MenuItem item)
            {
                switch (item.getItemId()) {
                    case R.id.share_whatsapp:
                        shareFavouritesInSocialMedia(text);
                        return true;
                    case R.id.export_pdf:
                        if (PermissionUtils.isStoragePermissionGranted(activity)) {
                            exportSongToPDF(text, favouriteService.findSongsByFavouriteName(text));
                        }

                    default:
                        return false;
                }
            }
        };
    }

    private void shareFavouritesInSocialMedia(String text)
    {
        Intent textShareIntent = new Intent(Intent.ACTION_SEND);
        textShareIntent.putExtra(Intent.EXTRA_TEXT, favouriteService.buildShareFavouriteFormat(text));
        textShareIntent.setType("text/plain");
        Intent intent = Intent.createChooser(textShareIntent, "Share with...");
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
    }
}

File: app/src/main/java/org/worshipsongs/service/PopupMenuService.kt
package org.worshipsongs.service

import android.annotation.TargetApi
import android.app.Activity
import android.content.Context.PRINT_SERVICE
import android.content.Intent
import android.graphics.Paint
import android.graphics.pdf.PdfDocument
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.print.PrintAttributes
import android.print.pdf.PrintedPdfDocument
import android.util.Log
import android.view.ContextThemeWrapper
import android.view.Gravity
import android.view.View
import android.widget.PopupMenu
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import org.apache.commons.lang3.StringUtils
import org.json.JSONArray
import org.json.JSONObject
import org.worshipsongs.BuildConfig
import org.worshipsongs.CommonConstants
import org.worshipsongs.R
import org.worshipsongs.WorshipSongApplication
import org.worshipsongs.WorshipSongApplication.Companion.context
import org.worshipsongs.activity.CustomYoutubeBoxActivity
import org.worshipsongs.activity.PresentSongActivity
import org.worshipsongs.dialog.FavouritesDialogFragment
import org.worshipsongs.domain.Song
import org.worshipsongs.utils.PermissionUtils
import java.io.File
import java.io.FileOutputStream
import java.util.*
import kotlin.collections.ArrayList

/**
 * author: Madasamy,Seenivasan, Vignesh Palanisamy
 * version: 1.0.0
 */

class PopupMenuService
{

    private val customTagColorService = CustomTagColorService()
    private val preferenceSettingService = UserPreferenceSettingService()
    private val favouriteService = FavouriteService()
    private val songService = SongService(context!!)
    private val databaseService = DatabaseService(context!!)
    private var songBookService = SongBookService(context!!)
    private val authorService = AuthorService(context!!)

    fun showPopupmenu(activity: AppCompatActivity, view: View, songName: String, hidePlay: Boolean)
    {
        val wrapper = ContextThemeWrapper(context, R.style.PopupMenu_Theme)
        val popupMenu: PopupMenu
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT)
        {
            popupMenu = PopupMenu(wrapper, view, Gravity.RIGHT)
        } else
        {
            popupMenu = PopupMenu(wrapper, view)
        }
        popupMenu.menuInflater.inflate(R.menu.favourite_share_option_menu, popupMenu.menu)
        val song = songService.findContentsByTitle(songName)
        val urlKey = song!!.urlKey
        val menuItem = popupMenu.menu.findItem(R.id.play_song)
        menuItem.isVisible = urlKey != null && urlKey.length > 0 && preferenceSettingService.isPlayVideo && hidePlay
        val exportMenuItem = popupMenu.menu.findItem(R.id.export_pdf)
        exportMenuItem.isVisible = android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT
        val presentSongMenuItem = popupMenu.menu.findItem(R.id.present_song)
        presentSongMenuItem.isVisible = false
        popupMenu.menu.findItem(R.id.export_OpenLPService).isVisible = false
        popupMenu.setOnMenuItemClickListener { item ->
            when (item.itemId)
            {
                R.id.addToList -> {
                    val bundle = Bundle()
                    bundle.putString(CommonConstants.TITLE_KEY, songName)
                    bundle.putString(CommonConstants.LOCALISED_TITLE_KEY, song.tamilTitle)
                    bundle.putInt(CommonConstants.ID, song.id)

                    val favouritesDialogFragment = FavouritesDialogFragment.newInstance(bundle)
                    favouritesDialogFragment.show(activity.supportFragmentManager, "FavouritesDialogFragment")
                    true
                }
                R.id.share_whatsapp ->
                {
                    shareSongInSocialMedia(songName, song)
                    true
                }
                R.id.play_song ->
                {
                    showYouTube(urlKey, songName)
                    true
                }
                R.id.present_song ->
                {
                    startPresentActivity(songName)
                    true
                }
                R.id.export_pdf ->
                {
                    if (PermissionUtils.isStoragePermissionGranted(activity))
                    {
                        exportSongToPDF(songName, Arrays.asList(song))
                    }
                    true
                }
                else -> false
            }
        }
        popupMenu.show()
    }

    private fun shareSongInSocialMedia(songName: String, song: Song)
    {
        val builder = StringBuilder()
        builder.append(songName).append("\n").append("\n")
        for (content in song.contents!!)
        {
            builder.append(customTagColorService.getFormattedLines(content))
            builder.append("\n")
        }
        builder.append(context!!.getString(R.string.share_info))
        Log.i(this@PopupMenuService.javaClass.simpleName, builder.toString())
        val textShareIntent = Intent(Intent.ACTION_SEND)
        textShareIntent.putExtra(Intent.EXTRA_TEXT, builder.toString())
        textShareIntent.type = "text/plain"
        val intent = Intent.createChooser(textShareIntent, "Share $songName with...")
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context!!.startActivity(intent)
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    private fun exportSongToPDF(songName: String, songs: List<Song>)
    {
        val file = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS), "$songName.pdf")

        val printAttrs = PrintAttributes.Builder().setColorMode(PrintAttributes.COLOR_MODE_COLOR).setMediaSize(PrintAttributes.MediaSize.ISO_A4).setResolution(PrintAttributes.Resolution("zooey", PRINT_SERVICE, 450, 700)).setMinMargins(PrintAttributes.Margins.NO_MARGINS).build()
        val document = PrintedPdfDocument(WorshipSongApplication.context!!, printAttrs)
        for (i in songs.indices)
        {
            val song = songs[i]
            val pageInfo = PdfDocument.PageInfo.Builder(450, 700, i).create()
            var page: PdfDocument.Page? = document.startPage(pageInfo)
            if (page != null)
            {
                val titleDesign = Paint()
                titleDesign.textAlign = Paint.Align.LEFT
                titleDesign.textSize = 18f
                val title = getTamilTitle(song) + song.title!!
                val titleLength = titleDesign.measureText(title)
                var yPos = 50f
                if (page.canvas.width > titleLength)
                {
                    val xPos = page.canvas.width / 2 - titleLength.toInt() / 2
                    page.canvas.drawText(title, xPos.toFloat(), 20f, titleDesign)
                } else
                {
                    var xPos = page.canvas.width / 2 - titleDesign.measureText(song.tamilTitle).toInt() / 2
                    page.canvas.drawText(song.tamilTitle!! + "/", xPos.toFloat(), 20f, titleDesign)
                    xPos = page.canvas.width / 2 - titleDesign.measureText(song.title).toInt() / 2
                    page.canvas.drawText(song.title!!, xPos.toFloat(), 45f, titleDesign)
                    yPos = 75f
                }
                for (content in song.contents!!)
                {
                    if (yPos > 620)
                    {
                        document.finishPage(page)
                        page = document.startPage(pageInfo)
                        yPos = 40f
                    }
                    yPos = customTagColorService.getFormattedPage(content, page!!, 10f, yPos)
                    yPos = yPos + 20
                }
            }
            document.finishPage(page)
        }
        try
        {
            val os = FileOutputStream(file)
            document.writeTo(os)
            document.close()
            os.close()
            val shareIntent = Intent(Intent.ACTION_SEND)
            shareIntent.type = "application/pdf"
            val uriForFile = FileProvider.getUriForFile(context!!, BuildConfig.APPLICATION_ID + ".provider", file)
            shareIntent.putExtra(Intent.EXTRA_STREAM, uriForFile)
            val intent = Intent.createChooser(shareIntent, "Share $songName with...")
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context!!.startActivity(intent)
            Log.i("done", file.absolutePath.toString())

        } catch (ex: Exception)
        {
            Log.e(PopupMenuService::class.java.simpleName, "Error occurred while exporting to PDF", ex)
        }

    }

    private fun getTamilTitle(song: Song): String
    {
        return if (StringUtils.isNotBlank(song.tamilTitle)) song.tamilTitle!! + "/" else ""
    }

    private fun showYouTube(urlKey: String?, songName: String)
    {
        Log.i(this.javaClass.simpleName, "Url key: " + urlKey!!)
        val youTubeIntent = Intent(context, CustomYoutubeBoxActivity::class.java)
        youTubeIntent.putExtra(CustomYoutubeBoxActivity.KEY_VIDEO_ID, urlKey)
        youTubeIntent.putExtra(CommonConstants.TITLE_KEY, songName)
        youTubeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context!!.startActivity(youTubeIntent)
    }

    private fun startPresentActivity(title: String)
    {
        val intent = Intent(context, PresentSongActivity::class.java)
        intent.putExtra(CommonConstants.TITLE_KEY, title)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context!!.startActivity(intent)
    }

    fun shareFavouritesInSocialMedia(activity: Activity, view: View, favouriteName: String)
    {
        val wrapper = ContextThemeWrapper(context, R.style.PopupMenu_Theme)
        val popupMenu: PopupMenu
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT)
        {
            popupMenu = PopupMenu(wrapper, view, Gravity.RIGHT)
        } else
        {
            popupMenu = PopupMenu(wrapper, view)
        }
        popupMenu.menuInflater.inflate(R.menu.favourite_share_option_menu, popupMenu.menu)
        popupMenu.menu.findItem(R.id.play_song).isVisible = false
        popupMenu.menu.findItem(R.id.present_song).isVisible = false
        popupMenu.menu.findItem(R.id.addToList).isVisible = false
        popupMenu.setOnMenuItemClickListener(getPopupMenuItemListener(activity, favouriteName))
        popupMenu.show()
    }

    private fun getPopupMenuItemListener(activity: Activity, favouriteName: String): PopupMenu.OnMenuItemClickListener
    {
        return PopupMenu.OnMenuItemClickListener { item ->
            when (item.itemId)
            {
                R.id.share_whatsapp ->
                {
                    shareFavouritesInSocialMedia(favouriteName)
                    true
                }
                R.id.export_pdf ->
                {
                    if (PermissionUtils.isStoragePermissionGranted(activity))
                    {
                        exportSongToPDF(favouriteName, favouriteService.findSongsByFavouriteName(favouriteName))
                    }
                    false
                }
                R.id.export_OpenLPService ->
                {
                    if (PermissionUtils.isStoragePermissionGranted(activity))
                    {
                        Toast.makeText(activity, "Exporting $favouriteName to service file...!", Toast.LENGTH_LONG).show()
                        shareAsOpenLPService(favouriteName)
                    }
                    true
                }
                else -> false
            }
        }
    }

    private fun shareAsOpenLPService(favouriteName: String)
    {
        val jsonArray = JSONArray()
        jsonArray.put(getGeneralInfo())

        for (song in favouriteService.findSongsByFavouriteName(favouriteName))
        {
            jsonArray.put(getServiceItems(song))
        }
        val serviceFile = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS), "$favouriteName.osj")

        File(serviceFile.toURI()).bufferedWriter().use { out ->
            out.write(jsonArray.toString())
        }
//        println("Json Array: $jsonArray")
    }

    private fun getGeneralInfo(): JSONObject
    {
        val openLPCoreInfo = HashMap<String, Any?>()
        openLPCoreInfo["lite-service"] = true
        openLPCoreInfo["service-theme"] = ""

        val generalInfo = HashMap<String, Any?>()
        generalInfo["openlp_core"] = openLPCoreInfo

        return JSONObject(generalInfo)
    }

    private fun getServiceItems(song: Song): JSONObject
    {
        val serviceItems = HashMap<String, Any?>()
        serviceItems["serviceitem"] = getServiceItem(song)
        return JSONObject(serviceItems)
    }

    private fun getServiceItem(song: Song): JSONObject
    {
        val serviceItem = HashMap<String, Any?>()
        serviceItem["header"] = getHeader(song)
        serviceItem["data"] = getDataElement(song)

//        println("serviceItem: $serviceItem")
        return JSONObject(serviceItem)
    }

    private fun getDataElement(song: Song): List<JSONObject>
    {
        val dataElements = arrayListOf<JSONObject>()

        val verseOrderList = song.verseOrder?.split(" ")?.toList()
        for(i in verseOrderList?.indices!!)
        {
            val rawSlide = song.contents?.get(i)
            val dataElement = HashMap<String, Any?>()
            dataElement["raw_slide"] = rawSlide
            dataElement["title"] = rawSlide?.substringBefore("{/y}")
            dataElement["verseTag"] = verseOrderList[i]
            dataElements.add(JSONObject(dataElement))
        }
        return dataElements
    }

    private fun getHeader(song: Song): JSONObject
    {
        val title = song.title
        val authors = authorService.findAuthorsByTitle(title)

        val headerElements = HashMap<String, Any?>()
        headerElements["start_time"] = 0
        headerElements["will_auto_start"] = false
        headerElements["type"] = 1
        headerElements["theme_overwritten"] = false
        headerElements["processor"] = null
        headerElements["from_plugin"] = false
        headerElements["auto_play_slides_loop"] = false
        headerElements["name"] = "songs"
        headerElements["notes"] = ""
        headerElements["capabilities"] = arrayListOf(2,1,5,8,9,13)
        headerElements["plugin"] = "songs"
        headerElements["background_audio"] = arrayListOf<Any>()
        headerElements["media_length"] = 0
        headerElements["icon"] = ":/plugins/plugin_songs.png"
        headerElements["search"] = ""
        headerElements["auto_play_slides_once"] = false
        headerElements["theme"] = null
        headerElements["timed_slide_interval"] = 0
        headerElements["end_time"] = 0

        headerElements["title"] = title
        headerElements["audit"] = arrayListOf(title, authors, "", "")
        headerElements["footer"] = arrayListOf(title, "Written by: "+ getWrittenBy(authors))
        headerElements["data"] = getHeaderData(title, authors)
        headerElements["xml_version"] = song.lyrics
        println(headerElements)
        return JSONObject(headerElements)
    }

    private fun getWrittenBy(authors: List<String>): String
    {
        if(authors.isNotEmpty()) {
            return when (authors.size) {
                1 -> { authors[0] }
                2 -> { authors[0] + " and " + authors[1] }
                else -> { authors.joinToString(separator = ", ") }
            }
        }
        return ""
    }

    private fun getHeaderData(title: String?, authors: List<String>): JSONObject
    {
        val headerData = HashMap<String, Any?>()
        headerData["title"] = title
        headerData["authors"] = authors.joinToString(separator = ", ")
        return JSONObject(headerData)
    }

    private fun shareFavouritesInSocialMedia(text: String)
    {
        val textShareIntent = Intent(Intent.ACTION_SEND)
        textShareIntent.putExtra(Intent.EXTRA_TEXT, favouriteService.buildShareFavouriteFormat(text))
        textShareIntent.type = "text/plain"
        val intent = Intent.createChooser(textShareIntent, "Share with...")
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context!!.startActivity(intent)
    }

    private fun getEnglishSongBookName(songBookNameList: List<String>): String
    {
        val songBookNames = ArrayList<String>()
        for(songBookName in songBookNameList) {
            songBookNames.add(databaseService.parseEnglishName(songBookName))
        }
        return songBookNames.joinToString()
    }

    private fun getTamilSongBookName(songBookNameList: List<String>): String
    {
        val songBookNames = ArrayList<String>()
        for(songBookName in songBookNameList) {
            songBookNames.add(databaseService.parseTamilName(songBookName))
        }
        return songBookNames.joinToString()
    }
}
